src/forcing/ForcingsEngine.cpp

```diff
@@ -0,0 +1,282 @@
+    const auto var_dim  = var_outputs_.size();
+
+    bmi_->Update();
+    time_current_index_++;
+    
+    const auto* ptr = static_cast<int*>(bmi_->GetValuePtr("CAT-ID"));
+    var_divides_ = std::vector<int>(ptr, ptr + id_dim);
+    var_cache_ = decltype(var_cache_){{ time_dim, id_dim, var_dim }};
+
+    // Cache initial iteration
+    const auto size = var_divides_.size();
+    for (size_type vi = 0; vi < var_outputs_.size(); vi++) {
+        // Get current values for each variable
+        const auto var_span = boost::span<const double>{
+            static_cast<double*>(bmi_->GetValuePtr(var_outputs_[vi])),
+            size
+        };
+
+        for (size_type di = 0; di < size; di++) {
+            // Set current variable value for each divide
+            var_cache_.at({{time_current_index_, di, vi}}) = var_span[di];
+        }
+    }
+
+    time_current_index_++;
+};
+
+ForcingsEngine::~ForcingsEngine()
+{
+    finalize();
+}
+
+ForcingsEngine& ForcingsEngine::instance(
+    const std::string& init,
+    std::size_t time_start,
+    std::size_t time_end
+)
+{
+    if (instances_[init] == nullptr) {
+        instances_[init] = std::make_unique<ForcingsEngine>(init, time_start, time_end);
+    }
+
+    return *instances_[init];
+}
+
+void ForcingsEngine::check_runtime_requirements()
+{
+    
+    // Check that the python module is installed.
+    {
+        auto interpreter_ = utils::ngenPy::InterpreterUtil::getInstance();
+        try {
+            auto mod = interpreter_->getModule("NextGen_Forcings_Engine");
+            auto cls = mod.attr("BMIForcingsEngine").cast<py::object>();
+        } catch(std::exception& e) {
+            throw std::runtime_error{
+                "Failed to initialize ForcingsEngine: ForcingsEngine python module is not installed or is not properly configured. (" + std::string{e.what()} + ")"
+            };
+        }
+    }
+
+    // Check that the WGRIB2 environment variable is defined
+    {
+        const auto* wgrib2_exec = std::getenv("WGRIB2");
+        if (wgrib2_exec == nullptr) {
+            throw std::runtime_error{"Failed to initialize ForcingsEngine: $WGRIB2 is not defined"};
+        }
+    }
+}
+
+boost::span<const std::string> ForcingsEngine::outputs() const noexcept
+{
+    return var_outputs_;
+}
+
+auto ForcingsEngine::time_begin() const noexcept
+  -> const clock_type::time_point&
+{
+    return time_start_;
+}
+
+auto ForcingsEngine::time_end() const noexcept
+  -> const clock_type::time_point&
+{
+    return time_end_;
+}
+
+auto ForcingsEngine::time_step() const noexcept
+  -> const clock_type::duration&
+{
+    return time_step_;
+}
+
+auto ForcingsEngine::time_index(time_t ctime) const noexcept
+  -> size_type
+{
+    const auto epoch = clock_type::from_time_t(ctime);
+    return time_index(epoch);
+}
+
+auto ForcingsEngine::time_index(clock_type::time_point epoch) const noexcept
+  -> size_type
+{
+    if (epoch < time_begin() || epoch > time_end()) {
```

Comment:
The way `time_dim` gets calculated, I think this may have to test `epoch >= time_end()` - i.e. it's a half-open interval, like C++'s usual `begin()/end()` iterator pairs

Reply:
I reverted this change -- I think the conditional logic may need revisiting. If we check for greater than or equal, then we'll get a `bad_index` value exactly at the end time; this happens in the `Timestepping` test, but I _think_ we'd want the last possible index if we call `get_ts_index_for_time` with the end time.

What may need to change is that if `epoch >= time_end()`, we always return the last possible index (i.e. in the case of the unit tests, that would be 5 for any time past `params.simulation_end_t`).

@PhilMiller Does that make sense to you? Or do you have a different suggestion?

Reply:
Before getting into the implementation details, is the interval spanning `time_begin(), time_end()` open or closed? I know there are actually deeper questions about the data semantics buried beneath this (e.g. instantaneous vs trailing average vs centered average vs etc), but I'm just wondering about implementation details. It seems implicit that queries at `time_begin()` are valid, and expected to produce whatever data. Are queries at `time_end()` valid, down the call stack? Do we expect such queries up the call stack?

I know there's some pieces of code here and in related codebases that do the indefinite extension of values past the end of the recorded observation/forecast period. I don't think we should immediately implement that here until we're very clear on whether we want it, whether it needs to be controlled in some fashion, and whether there are distinctions between requests past `time_end()` that are valid and those that are simply erroneous.

Reply:
Following the other data providers, it should be a half-open interval, i.e. `[time_begin(), time_end)`. However, I think we want to have an index at `time_end`, because, I _think_, the semantic of `bmi.UpdateUntil(time_end)` is consistent with the half-open interval (i.e. we can update up to the end time, but not past it).

As far as the forcings engine is concerned, it will update until while its current time value is less than whatever sent to it, as long as the necessary GRIB files are available.

It probably makes sense _not_ to extend the the index to invalid times, so I agree with that last part.